File: packages/honker-ex/examples/atomic.exs
# Atomic business-write + enqueue in one transaction.
#
#   mix run examples/atomic.exs

ext_path =
  System.get_env("HONKER_EXTENSION_PATH") ||
    Path.expand("../../../target/release/libhonker_ext.dylib", __DIR__)

dir = Path.join(System.tmp_dir!(), "honker-#{System.unique_integer([:positive])}")
File.mkdir_p!(dir)

try do
  {:ok, db} = Honker.open(Path.join(dir, "app.db"), extension_path: ext_path)

  {:ok, _} = Honker.query_first(db.conn,
    "CREATE TABLE orders (id INTEGER PRIMARY KEY, user_id INTEGER, total INTEGER)", [])

  count = fn sql ->
    {:ok, [n]} = Honker.query_first(db.conn, sql, [])
    n
  end

  # Success path.
  {:ok, _} = Honker.query_first(db.conn, "BEGIN IMMEDIATE", [])
  {:ok, _} = Honker.query_first(db.conn,
    "INSERT INTO orders (user_id, total) VALUES (?1, ?2)", [42, 9900])
  {:ok, _} = Honker.query_first(db.conn,
    "SELECT honker_enqueue(?1, ?2, ?3, ?4, ?5, ?6, ?7)",
    ["emails",
     Jason.encode!(%{"to" => "alice@example.com", "order_id" => 42}),
     nil, nil, 0, 3, nil])
  {:ok, _} = Honker.query_first(db.conn, "COMMIT", [])

  IO.puts(
    "committed: #{count.("SELECT COUNT(*) FROM orders")} order(s), " <>
    "#{count.("SELECT COUNT(*) FROM _honker_live WHERE queue='emails'")} job(s)"
  )

  # Rollback path.
  {:ok, _} = Honker.query_first(db.conn, "BEGIN IMMEDIATE", [])
  {:ok, _} = Honker.query_first(db.conn,
    "INSERT INTO orders (user_id, total) VALUES (?1, ?2)", [43, 5000])
  {:ok, _} = Honker.query_first(db.conn,
    "SELECT honker_enqueue(?1, ?2, ?3, ?4, ?5, ?6, ?7)",
    ["emails",
     Jason.encode!(%{"to" => "bob@example.com", "order_id" => 43}),
     nil, nil, 0, 3, nil])
  {:ok, _} = Honker.query_first(db.conn, "ROLLBACK", [])
  IO.puts("rolled back: simulated payment-processing failure")

  IO.puts(
    "after rollback: #{count.("SELECT COUNT(*) FROM orders")} order(s), " <>
    "#{count.("SELECT COUNT(*) FROM _honker_live WHERE queue='emails'")} job(s)"
  )
  IO.puts("atomic enqueue + rollback both work as expected")
after
  File.rm_rf!(dir)
end

File: packages/honker-bun/examples/atomic.ts
// Atomic business-write + enqueue in one transaction.
//
//   bun run examples/atomic.ts

import { open } from "../src/index.ts";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const EXT_PATH =
  process.env.HONKER_EXTENSION_PATH ??
  new URL(
    "../../../target/release/libhonker_ext.dylib",
    import.meta.url,
  ).pathname;

const dir = mkdtempSync(join(tmpdir(), "honker-"));
const db = open(join(dir, "app.db"), EXT_PATH);

try {
  db.raw.exec(
    "CREATE TABLE orders (id INTEGER PRIMARY KEY, user_id INTEGER, total INTEGER)",
  );

  // Success path.
  db.raw.transaction(() => {
    db.raw.run(
      "INSERT INTO orders (user_id, total) VALUES (?, ?)",
      [42, 9900],
    );
    db.raw.run(
      "SELECT honker_enqueue(?, ?, ?, ?, ?, ?, ?)",
      [
        "emails",
        JSON.stringify({ to: "alice@example.com", order_id: 42 }),
        null, null, 0, 3, null,
      ],
    );
  })();

  const count = (sql: string) =>
    (db.raw.query<{ c: number }, []>(sql).get()?.c ?? 0);
  console.log(`committed: ${count("SELECT COUNT(*) AS c FROM orders")} order(s), ` +
              `${count("SELECT COUNT(*) AS c FROM _honker_live WHERE queue='emails'")} job(s)`);

  // Rollback path — bun:sqlite rollbacks on any thrown error inside the tx.
  try {
    db.raw.transaction(() => {
      db.raw.run(
        "INSERT INTO orders (user_id, total) VALUES (?, ?)",
        [43, 5000],
      );
      db.raw.run(
        "SELECT honker_enqueue(?, ?, ?, ?, ?, ?, ?)",
        [
          "emails",
          JSON.stringify({ to: "bob@example.com", order_id: 43 }),
          null, null, 0, 3, null,
        ],
      );
      throw new Error("boom — simulated payment-processing failure");
    })();
  } catch (e) {
    console.log(`rolled back: ${(e as Error).message}`);
  }

  console.log(`after rollback: ${count("SELECT COUNT(*) AS c FROM orders")} order(s), ` +
              `${count("SELECT COUNT(*) AS c FROM _honker_live WHERE queue='emails'")} job(s)`);
  console.log("atomic enqueue + rollback both work as expected");
} finally {
  db.close();
  rmSync(dir, { recursive: true, force: true });
}
